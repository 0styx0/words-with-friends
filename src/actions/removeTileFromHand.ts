import types from './types';
import Tile from '../interfaces/Tile';
import PlayerClass from '../classes/Player';

export default function removeTileFromHand(Players: ReadonlyArray<Readonly<PlayerClass>>, tile: Tile) {
    return {
        type: types.REMOVE_TILE_FROM_HAND,
        Players,
        tile
    };
}