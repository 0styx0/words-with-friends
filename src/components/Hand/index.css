.tileHand div {
    height: 6vh; /* all tile holders */
}
