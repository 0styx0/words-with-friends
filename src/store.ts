import reducers from './reducers';
import { createStore, compose } from 'redux';
import Player from './classes/Player';
import Tilebag from './classes/Tilebag';
import Board from './classes/Board';

const enhancers = compose('devToolsExtension' in window ?
    (window as {} as { devToolsExtension: Function }).devToolsExtension!() :
    (f: {}) => f);

export const defaultState = {
    turn: 0,
    Players: [new Player(true, 0), new Player(false, 1)], // turns switched around on INIT_PLAYERS
    board: new Board(),
    Tilebag: new Tilebag()
};

const store = createStore(reducers, defaultState, enhancers);

export default store;
