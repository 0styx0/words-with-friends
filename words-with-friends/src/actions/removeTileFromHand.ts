import types from './types';
import Tile from '../interfaces/Tile';
import PlayerClass from '../classes/Player';

export default function removeTileFromHand(Player: PlayerClass, tile: Tile) {
    return {
        type: types.REMOVE_TILE_FROM_HAND,
        Player,
        tile
    };
}