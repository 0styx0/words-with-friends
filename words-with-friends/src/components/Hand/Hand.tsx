import * as React from 'react';
import './index.css';
import Tilebag from '../../classes/Tilebag';
import Tile from '../../interfaces/Tile';
import Hand from './';


interface State {
    tiles: Tile[];
    removedTiles: Tile[];
}

interface Props {
    className?: string;
    canDrag: boolean;
}

export default class HandContainer extends React.Component<Props, State> {

    constructor() {
        super();

        this.state = {
            tiles: [],
            removedTiles: []
        };
    }

    componentDidMount() {

        this.generateTiles();
    }

    componentWillReceiveProps() {

        const tiles = [...this.state.tiles];

        // remove tiles that were put on the board
        this.state.removedTiles.map(tile => {

            const indexOfTile = tiles.indexOf(tile);
            tiles.splice(indexOfTile, 1);
        });

        this.generateTiles(tiles); // after a turn, canDrag is updated so latching onto that
    }

    /**
     * Sets state so there is a full hand of Tiles
     */
    generateTiles(tiles: Tile[] = []) {

        while (tiles.length < 7 && Tilebag.tiles.length > 0) {
            tiles.push(Tilebag.getRandomTile());
        }

        this.setState({
           tiles: tiles,
           removedTiles: []
        });
    }

    render() {

        return (
            <Hand
              key={+this.props.canDrag}
              tiles={this.state.tiles}
              canDrag={this.props.canDrag}
              className={this.props.className}
              removeTile={(tile: Tile) => {
                  this.setState({
                      removedTiles: this.state.removedTiles.concat([tile])
                  });
              }}
            />
        );
    }
}