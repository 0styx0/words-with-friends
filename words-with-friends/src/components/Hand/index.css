.tileHand { /* wrapper */
    width: 10vw;
    position: absolute;
    bottom: 61vh;
    top: 8vh;
}

.rightHand {
    right: 0;
}

.tileHand div {
    height: 6vh; /* all tile holders */
}
