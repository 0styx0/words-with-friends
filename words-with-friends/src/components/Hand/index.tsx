import * as React from 'react';
import TileHolder from '../TileHolder/TileHolder';
import './index.css';
import Tile from '../../interfaces/Tile';

interface Props {
    className?: string;
    tiles: Tile[];
}

/**
 * A player's hand
 */
export default function Hand(props: Props) {

    const handHolders = props.tiles.map((tile, i) => (
        <TileHolder tile={tile} key={i + tile.letter} {...[] as any}/>
    ));

    return <div className="tileHand">{handHolders}</div>;
}