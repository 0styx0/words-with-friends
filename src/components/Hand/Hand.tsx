import * as React from 'react';
import './index.css';
import HandTileHolder from '../TileHolder/HandTileHolder';
import { bindActionCreators } from 'redux';
import { connect, Dispatch } from 'react-redux';
import actionCreators from '../../actions';
import { defaultState } from '../../store';
import Tile from '../../interfaces/Tile';
import Board from '../../classes/Board';


function mapStateToProps(state: typeof defaultState, props: {playerIndex: number}) {

    return {
        tiles: state.Players[props.playerIndex].tiles,
        board: state.board,
        tileNumber: state.Players[props.playerIndex].tiles.length, // force refresh after tile is removed
        turn: state.turn // force refresh after turn
    };
}

function mapDispatchToProps(dispatch: Dispatch<typeof defaultState>) {
    return bindActionCreators(actionCreators, dispatch);
}

type Props = typeof actionCreators & typeof defaultState & {
    tiles: Tile[],
    board: Board,
    tileNumber: number,
    turn: number
};

function HandContainer(props: Readonly<Props>) {

    const tiles = props.tiles;

    const handHolders = tiles.map((tile, i) => (
        <HandTileHolder tile={tile} key={i + tile.letter} {...[] as any} />
    ));

    const lengthOfScrabbleHand = 7;

    while (handHolders.length < lengthOfScrabbleHand) {
        handHolders.push(<HandTileHolder key={handHolders.length} {...[] as any} />);
    }

    return <div className="tileHand">{handHolders}</div>;
}

export default connect(mapStateToProps, mapDispatchToProps)(HandContainer as any);
