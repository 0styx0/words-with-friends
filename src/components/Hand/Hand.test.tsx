import * as React from 'react';
import Hand from './Hand';
import { Provider } from 'react-redux';
import store, { defaultState } from '../../store';
import * as renderer from 'react-test-renderer';
import removeTileFromHand from '../../actions/removeTileFromHand';


describe('<Hand />', () => {

    describe('renders 7 <HandTileHolder when Player has', () => {

        it('7 (max) tiles', () => {

            const tree = renderer.create(
                <Provider store={store}>
                    <Hand {...[] as any} playerIndex={0} />
                </Provider>
            ).toJSON();

            expect(tree).toMatchSnapshot();
        });

        it('less than 7 (max) tiles', () => {

            const state = store.getState() as typeof defaultState;

            store.dispatch(removeTileFromHand(state.Players, state.Players[0].tiles[0]));

            const tree = renderer.create(
                <Provider store={store}>
                    <Hand {...[] as any} playerIndex={0} />
                </Provider>
            ).toJSON();

            expect(tree).toMatchSnapshot();
        });
    });

});