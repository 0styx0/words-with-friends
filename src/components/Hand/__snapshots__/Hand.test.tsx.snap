// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Hand /> renders 7 <HandTileHolder when Player has 7 (max) tiles 1`] = `
<div
  className="tileHand"
>
  <div
    className="tileHolder"
    onDragOver={[Function]}
    onDrop={[Function]}
  >
    <span>
      
    </span>
  </div>
  <div
    className="tileHolder"
    onDragOver={[Function]}
    onDrop={[Function]}
  >
    <span>
      
    </span>
  </div>
  <div
    className="tileHolder"
    onDragOver={[Function]}
    onDrop={[Function]}
  >
    <span>
      
    </span>
  </div>
  <div
    className="tileHolder"
    onDragOver={[Function]}
    onDrop={[Function]}
  >
    <span>
      
    </span>
  </div>
  <div
    className="tileHolder"
    onDragOver={[Function]}
    onDrop={[Function]}
  >
    <span>
      
    </span>
  </div>
  <div
    className="tileHolder"
    onDragOver={[Function]}
    onDrop={[Function]}
  >
    <span>
      
    </span>
  </div>
  <div
    className="tileHolder"
    onDragOver={[Function]}
    onDrop={[Function]}
  >
    <span>
      
    </span>
  </div>
</div>
`;

exports[`<Hand /> renders 7 <HandTileHolder when Player has less than 7 (max) tiles 1`] = `
<div
  className="tileHand"
>
  <div
    className="tileHolder"
    onDragOver={[Function]}
    onDrop={[Function]}
  >
    <span>
      
    </span>
  </div>
  <div
    className="tileHolder"
    onDragOver={[Function]}
    onDrop={[Function]}
  >
    <span>
      
    </span>
  </div>
  <div
    className="tileHolder"
    onDragOver={[Function]}
    onDrop={[Function]}
  >
    <span>
      
    </span>
  </div>
  <div
    className="tileHolder"
    onDragOver={[Function]}
    onDrop={[Function]}
  >
    <span>
      
    </span>
  </div>
  <div
    className="tileHolder"
    onDragOver={[Function]}
    onDrop={[Function]}
  >
    <span>
      
    </span>
  </div>
  <div
    className="tileHolder"
    onDragOver={[Function]}
    onDrop={[Function]}
  >
    <span>
      
    </span>
  </div>
  <div
    className="tileHolder"
    onDragOver={[Function]}
    onDrop={[Function]}
  >
    <span>
      
    </span>
  </div>
</div>
`;
